### PR TITLE
Add Spring Cloud in Azure Explorer

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/IDataRefreshableComponent.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.tooling.msservices.serviceexplorer;
+
+public interface IDataRefreshableComponent<T, U> {
+    void noDataRefresh(T mainData, U additionalData);
```
Comment (Flanker32) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/IDataRefreshableComponent.java]: notifyDataRefresh?

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/springcloud/SpringCloudNodePresenter.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SpringCloudNodePresenter<V extends SpringCloudNodeView> extends MvpPresenter<V> {
+    public void onRefreshSpringCloudServiceNode(final String subscriptionId, final String clusterId)
+            throws IOException {
+        final SpringCloudNodeView view = getMvpView();
+        if (view != null) {
+            final List<AppResourceInner> appList = AzureSpringCloudMvpModel.listAppsByClusterId(clusterId);
+            final Observable<DeploymentResource> deployList = AzureSpringCloudMvpModel.listAllDeploymentsByClusterId(clusterId);
+            final Map<String, DeploymentResource> activeDeployments = new ConcurrentHashMap<>();
```
Comment (Flanker32) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/springcloud/SpringCloudNodePresenter.java]: is there any multi-thread issue  we need to use `ConcurrentHashMap`

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/IDataRefreshableComponent.java (hunk 1 above)
Comment (andxu) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/IDataRefreshableComponent.java]: fixed in https://github.com/microsoft/azure-tools-for-java/pull/4201/commits/f1587a5d775156e20c57f10db0f3fe654c1533b4

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/springcloud/SpringCloudNodePresenter.java (hunk 2 above)
Comment (andxu) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/springcloud/SpringCloudNodePresenter.java]: Fixed in https://github.com/microsoft/azure-tools-for-java/pull/4201/commits/4131793daf4006c6d1fe86e377af071ea304bfd3

**Comment by andxu** [comment]
replaced with https://github.com/microsoft/azure-tools-for-java/pull/4206

**Pull request closed without merging**